### DISCUSSION TITLE
Duplicate "graphql" modules cannot be used at the same time since different versions...

### DISCUSSION THREAD

**Discussion by Stupidism** [payloadcms/payload#1856] [Q&A]
# Bug Report

I'm using payload in a nx monorepo, I met this error after upgrading payload to the newest version.  And I found it [was upgraded](https://github.com/payloadcms/payload/commit/57f5f5ec439b5aee1d46bff0bf31aac6148f16b2) in 1.4.0.

```
nx run seo-payload:serve

 >  NX   USING REMOTE CACHE


> nx run seo-payload:serve

chunk (runtime: config) config.js (config) 283 KiB [entry] [rendered]
chunk (runtime: main) main.js (main) 11.7 KiB [entry] [rendered]
webpack compiled successfully (1799b20c209ec786)
Debugger listening on ws://localhost:9229/20d445a8-52d6-422f-9523-b6edace32404
Debugger listening on ws://localhost:9229/20d445a8-52d6-422f-9523-b6edace32404
For help, see: https://nodejs.org/en/docs/inspector
body-parser deprecated undefined extended: provide extended option dist/apps/seo-payload/main.js:388:27
[02:01:19] INFO (payload): Starting Payload...

/Users/stupidism/Documents/jerry-serverless/node_modules/payload/node_modules/graphql/jsutils/instanceOf.js:44
            throw new Error(`Cannot use ${className} "${stringifiedValue}" from another module or realm.
                  ^
Error: Cannot use GraphQLScalarType "DateTime" from another module or realm.

Ensure that there is only one instance of "graphql" in the node_modules
directory. If different versions of "graphql" are the dependencies of other
relied on modules, use "resolutions" to ensure only one version is installed.

https://yarnpkg.com/en/docs/selective-version-resolutions

Duplicate "graphql" modules cannot be used at the same time since different
versions may have different capabilities and behavior. The data from one
version used in the function from another could produce confusing and
spurious results.
    at instanceOf (/Users/stupidism/Documents/jerry-serverless/node_modules/payload/node_modules/graphql/jsutils/instanceOf.js:44:19)
    at isScalarType (/Users/stupidism/Documents/jerry-serverless/node_modules/payload/node_modules/graphql/type/definition.js:117:37)
    at isType (/Users/stupidism/Documents/jerry-serverless/node_modules/payload/node_modules/graphql/type/definition.js:92:5)
    at isNullableType (/Users/stupidism/Documents/jerry-serverless/node_modules/payload/node_modules/graphql/type/definition.js:440:10)
    at new GraphQLNonNull (/Users/stupidism/Documents/jerry-serverless/node_modules/payload/node_modules/graphql/type/definition.js:394:5)
    at /Users/stupidism/Documents/jerry-serverless/node_modules/payload/dist/collections/graphql/init.js:101:23
    at Array.forEach (<anonymous>)
    at initCollectionsGraphQL (/Users/stupidism/Documents/jerry-serverless/node_modules/payload/dist/collections/graphql/init.js:60:38)
    at registerSchema (/Users/stupidism/Documents/jerry-serverless/node_modules/payload/dist/graphql/registerSchema.js:58:24)
    at init (/Users/stupidism/Documents/jerry-serverless/node_modules/payload/dist/init.js:79:38)
^C%
```

## Steps to Reproduce

https://github.com/Stupidism/nx-payload-demo.

1. `git clone git@github.com:Stupidism/nx-payload-demo.git`
2. `PUPPETEER_SKIP_DOWNLOAD=1 CYPRESS_INSTALL_BINARY=0 npm install`
3. `nx run seo-payload:serve`

```
 >  NX   USING REMOTE CACHE


> nx run seo-payload:serve

chunk (runtime: config) config.js (config) 6.7 KiB [entry] [rendered]
chunk (runtime: main) main.js (main) 2.85 KiB [entry] [rendered]
webpack compiled successfully (c07d01079b667599)
Debugger listening on ws://localhost:9229/2a44fcbc-77c4-42f7-9399-19de2f2197dd
Debugger listening on ws://localhost:9229/2a44fcbc-77c4-42f7-9399-19de2f2197dd
For help, see: https://nodejs.org/en/docs/inspector
body-parser deprecated undefined extended: provide extended option dist/apps/seo-payload/main.js:115:27
[13:36:20] INFO (payload): Starting Payload...

/Users/stupidism/Documents/demos/nx-payload-demo/node_modules/payload/node_modules/graphql/jsutils/instanceOf.js:44
            throw new Error(`Cannot use ${className} "${stringifiedValue}" from another module or realm.
                  ^
Error: Cannot use GraphQLScalarType "DateTime" from another module or realm.

Ensure that there is only one instance of "graphql" in the node_modules
directory. If different versions of "graphql" are the dependencies of other
relied on modules, use "resolutions" to ensure only one version is installed.

https://yarnpkg.com/en/docs/selective-version-resolutions

Duplicate "graphql" modules cannot be used at the same time since different
versions may have different capabilities and behavior. The data from one
version used in the function from another could produce confusing and
spurious results.
    at instanceOf (/Users/stupidism/Documents/demos/nx-payload-demo/node_modules/payload/node_modules/graphql/jsutils/instanceOf.js:44:19)
    at isScalarType (/Users/stupidism/Documents/demos/nx-payload-demo/node_modules/payload/node_modules/graphql/type/definition.js:117:37)
    at isType (/Users/stupidism/Documents/demos/nx-payload-demo/node_modules/payload/node_modules/graphql/type/definition.js:92:5)
    at isNullableType (/Users/stupidism/Documents/demos/nx-payload-demo/node_modules/payload/node_modules/graphql/type/definition.js:440:10)
    at new GraphQLNonNull (/Users/stupidism/Documents/demos/nx-payload-demo/node_modules/payload/node_modules/graphql/type/definition.js:394:5)
    at /Users/stupidism/Documents/demos/nx-payload-demo/node_modules/payload/src/collections/graphql/init.ts:96:15
    at Array.forEach (<anonymous>)
    at initCollectionsGraphQL (/Users/stupidism/Documents/demos/nx-payload-demo/node_modules/payload/src/collections/graphql/init.ts:42:36)
    at registerSchema (/Users/stupidism/Documents/demos/nx-payload-demo/node_modules/payload/src/graphql/registerSchema.ts:36:18)
    at init (/Users/stupidism/Documents/demos/nx-payload-demo/node_modules/payload/src/init.ts:89:19)

```

**Comment by jmikrut** [comment]
Hey @Stupidism — this is actually not a bug with Payload and more of a restriction with GraphQL itself.

I assume that this is related to a custom GraphQL mutation or query that you have written. If you import types directly from a GraphQL dependency, this can arise. 

Notice in our docs how we provide `GraphQL` as an argument to custom queries and mutations - _**this is why.**_  This is literally the same exact GraphQL dependency that Payload uses, so if you use types coming from that GraphQL argument, you will not see this issue.

Otherwise, if you follow the suggested GraphQL solution in the error message, and specify a `resolutions` to GraphQL in your `package.json`, that should allow you to dodge this issue as well.

I will convert this to a discussion and we can keep convo going there. But this is expected and there is little that Payload can do to rectify this, unfortunately. Got me a few times myself when having multiple copies / versions of GraphQL installed.

**Reply by Stupidism** [reply]
But actually I didn't import any graphql types directly or write any custom queries or mutations.

I found this library very suspicious in the dependency list: `express-graphql` and they actuall have a issue about supporting graphql 16:

https://github.com/graphql/express-graphql/issues/797


And since the library is deprecated, some of the repliers say they switched to another library `graphql-helix`: https://github.com/graphql/express-graphql/issues/797#issuecomment-1131300228


PS: I'm trying to create a repro repo.

**Reply by TimHal** [reply]
I would argue this is at least partially a payload-related bug as this breaks the npm-based installation. 
NPM does come with overrides https://github.com/npm/cli/releases/tag/v8.3.0 but simply installing/updating payload as before is not possible anymore.

Workaround: 
 - delete `node_modules`
 - install with yarn

**Reply by falko100** [reply]
I'm having the same issue on npm. Works fine on yarn, but my entire CI/CD is done with npm..
Is there a solution for this?

**Reply by sdjnes** [reply]
As @TimHal said, I also think it's partially a payload bug.

As far as GraphQL goes, I haven't touched it at all in my current Payload setup (still setting collections etc up). This particular issue shows up for me when I am running my Payload app in a Docker container.

I'm using `npm` and can solve it with @crocq1's suggestion of using `overrides`. 

If I run `npm ls graphql`, I can see the mismatch in versions is coming Payload and its dependencies. This is the output when `overrides` is set vs when it's not:

```diff
    `-- payload@1.6.22
       +-- express-graphql@0.12.0
       | `-- graphql@15.8.0 deduped
       +-- graphql-query-complexity@0.12.0
       | `-- graphql@15.8.0 deduped
       +-- graphql-scalars@1.20.4
       | `-- graphql@15.8.0 deduped
       +-- graphql-type-json@0.3.2
       | `-- graphql@15.8.0 deduped
-  `-- graphql@16.6.0
+  `-- graphql@15.8.0 overridden
```

I believe Payload either needs to upgrade the dependencies that are using 15.8.0 (so they use 16.6.0) or Payload's graphql dependency needs to be downgraded to `15.8.0` to match

**Comment by sifferhans** [comment]
I have the same problem. I tried using Payload in a monorepo together with a Nuxt app, and I get this error. I haven't installed or used anything graphql-related in the Nuxt app, so I'm quite confused

**Comment by Stupidism** [comment]
@jmikrut Sorry for the late response. Here's the repro repo: https://github.com/Stupidism/nx-payload-demo. Can you take a look? This is purely an empty project but with the dependencies from my real project.

**Reply by DanRibbens** [reply]
I reproduced it from your steps above. What confuses me is why is this package manager specific? Either way, it does seem we need to investigate replacing `express-graphql` at some point.

**Reply by Stupidism** [reply]
Thank you @DanRibbens . It's a huge mono repo. We have about 50 apps and 50 libs(I only contribute to 3 apps and 5 libs). The dependencies are complicated too as you can tell from the package.json. I don't think we can switch to yarn or pnpm with low cost.

**Reply by Stupidism** [reply]
I think @crocq1 's [workaround](https://github.com/payloadcms/payload/discussions/1856#discussioncomment-4919054) already helped me with this issue. But the new swc issue is killing me. 

@DanRibbens Do you have any idea why swc-loader has this error? Is it also a version mismatch for swc?

**Comment by crocq1** [comment]
Do **npm ls graphql** to see the different versions.

Then try to add overrides in package.json (https://docs.npmjs.com/cli/v8/configuring-npm/package-json#overrides)

For example I did this : 

`"overrides": {
  "graphql@>15.7.0 <16.7.0": "^15.8.0"
}`

**Reply by Stupidism** [reply]
I tried this solution but it didn't work.

**Reply by Stupidism** [reply]
The result of npm ls graphql is:

```
0 verbose cli /Users/stupidism/.nvm/versions/node/v18.13.0/bin/node /Users/stupidism/.nvm/versions/node/v18.13.0/bin/npm
1 info using npm@8.19.3
2 info using node@v18.13.0
3 timing npm:load:whichnode Completed in 0ms
4 timing config:load:defaults Completed in 1ms
5 timing config:load:file:/Users/stupidism/.nvm/versions/node/v18.13.0/lib/node_modules/npm/npmrc Completed in 0ms
6 timing config:load:builtin Completed in 0ms
7 timing config:load:cli Completed in 1ms
8 timing config:load:env Completed in 0ms
9 timing config:load:file:/Users/stupidism/Documents/demos/nx-payload-demo/.npmrc Completed in 2ms
10 timing config:load:project Completed in 7ms
11 timing config:load:file:/Users/stupidism/.npmrc Completed in 0ms
12 timing config:load:user Completed in 0ms
13 timing config:load:file:/Users/stupidism/.nvm/versions/node/v18.13.0/etc/npmrc Completed in 0ms
14 timing config:load:global Completed in 0ms
15 timing config:load:validate Completed in 1ms
16 timing config:load:credentials Completed in 0ms
17 timing config:load:setEnvs Completed in 1ms
18 timing config:load Completed in 11ms
19 timing npm:load:configload Completed in 11ms
20 timing npm:load:mkdirpcache Completed in 0ms
21 timing npm:load:mkdirplogs Completed in 0ms
22 verbose title npm ls graphql error.log
23 verbose argv "ls" "graphql" "error.log"
24 timing npm:load:setTitle Completed in 7ms
25 timing config:load:flatten Completed in 1ms
26 timing npm:load:display Completed in 2ms
27 verbose logfile logs-max:10 dir:/Users/stupidism/.npm/_logs
28 verbose logfile /Users/stupidism/.npm/_logs/2023-02-09T13_53_59_850Z-debug-0.log
29 timing npm:load:logFile Completed in 4ms
30 timing npm:load:timers Completed in 0ms
31 timing npm:load:configScope Completed in 0ms
32 timing npm:load Completed in 24ms
33 timing arborist:ctor Completed in 0ms
34 silly logfile start cleaning logs, removing 1 files
35 silly logfile done cleaning log files
36 verbose shrinkwrap failed to load node_modules/.package-lock.json missing from node_modules: @types/jest-puppeteer
37 timing command:ls Completed in 5403ms
38 verbose stack Error: invalid: graphql@15.4.0 /Users/stupidism/Documents/demos/nx-payload-demo/node_modules/graphql
38 verbose stack     at LS.exec (/Users/stupidism/.nvm/versions/node/v18.13.0/lib/node_modules/npm/lib/commands/ls.js:205:9)
38 verbose stack     at async module.exports (/Users/stupidism/.nvm/versions/node/v18.13.0/lib/node_modules/npm/lib/cli.js:78:5)
39 verbose cwd /Users/stupidism/Documents/demos/nx-payload-demo
40 verbose Darwin 21.6.0
41 verbose node v18.13.0
42 verbose npm  v8.19.3
43 error code ELSPROBLEMS
44 error invalid: graphql@15.4.0 /Users/stupidism/Documents/demos/nx-payload-demo/node_modules/graphql
45 verbose exit 1
46 timing npm Completed in 5525ms
47 verbose code 1
48 error A complete log of this run can be found in:
48 error     /Users/stupidism/.npm/_logs/2023-02-09T13_53_59_850Z-debug-0.log

```

**Reply by crocq1** [reply]
Try to delete node_modules & package-lock.json and then npm install
Then retry npm ls graphql

You should see something like this : 

npm ls graphql
backend-payload@1.0.0 C:\projects\backend-payload
└─┬ payload@1.6.7
  ├─┬ express-graphql@0.12.0
  │ └── graphql@16.6.0 deduped
  ├─┬ graphql-query-complexity@0.12.0
  │ └── graphql@16.6.0 deduped
  ├─┬ graphql-scalars@1.20.1
  │ └── graphql@16.6.0 deduped
  ├─┬ graphql-type-json@0.3.2
  │ └── graphql@16.6.0 deduped
  └── graphql@16.6.0

**Reply by Stupidism** [reply]
@crocq1 I tried your workaround and get passed the original issue, but I met this issue instead:

> The updated commit has been pushed to the reproduce repo.

```
ERROR in ./dist/apps/seo-payload/config.js
Module build failed (from ./node_modules/swc-loader/src/index.js):
Error: unknown field `presets` at line 1 column 400

webpack compiled with 1 error

```

**Comment by sifferhans** [comment]
I migrated to [pnpm](https://pnpm.io) and did `pnpm install --shamefully-hoist` and I think that was the solution for me:)

**Comment by Stupidism** [comment]
I believe above answers about `yarn` or `pnpm` can solve this. But **I'm in a very huge monorepo, I cannot switch to another package manager**. Can anybody provide a workaround with `npm`?

**Reply by sdjnes** [reply]
Are you able to downgrade the npm version? I have the same issue with Payload with node 16 and higher, but node 14 works

**Comment by DanRibbens** [comment]
I finally made time to address the express-graphql dependency!

https://github.com/payloadcms/payload/pull/2484

We're working to review and release this soon.

**Comment by blueMountain007** [comment]
@DanRibbens  I encountered the same problem when upgrading from v1.11.1 to v2.3.1. Is there any workaround? Thanks!
In addition, I use npm

**Reply by DanRibbens** [reply]
Hey @xin-ze, 
We've seen this come up and go away from time to time. I would use `npm why graphql` `rm -rf node_modules` and any lock files then install again. If that doesn't fix it you can try and use `yarn why` to figure out what dependencies are causing this divergence.

You can get around it with resolutions most of the time, but it isn't ideal to solve it this way.

Feel free to open a new issue with your package manager and we can look into it more.
